### PR TITLE
 resource attributes layer (refactoring)

### PR DESCRIPTION
**Pull request by Fivell**
introducing correct handling foreign keys attributes on resource level.

Problems:

https://github.com/activeadmin/activeadmin/issues/4298#issuecomment-185467400 content_columns hardcoded in many places 

https://github.com/activeadmin/activeadmin/commit/f21cff30e7212706a39710e2c70ad0af8087ccea see Notes to commit message, example when renaming foreign key affects default show content

https://github.com/activeadmin/activeadmin/issues/4298#issuecomment-185849747  ActiveAdmin can't figure out if column is foreign key or not

https://github.com/activeadmin/activeadmin/commit/3efb359533df46e76bb4355a02ad31e727fb0da3#diff-b2d41db115eac24dba6f5e64bda8a690R53 introduced reflection parsing for each find_value method call because we still not know if this foreign key or not.

Idea is to initialise attributes for each resource and skip heavy magic in runtime 

@timoschilling , @varyonic ,  @deivid-rodriguez  before continue working on this, I would like to have feedback

### CODE REVIEW

**Comment by Fivell** [comment]
@timoschilling , @varyonic , @deivid-rodriguez, specs fixed, ready for review

**Comment by Fivell** [comment]
@timoschilling , any update ?

File: lib/active_admin/resource.rb [hunk 1]
```diff
@@ -162,6 +165,22 @@ def find_resource(id)
...
+    def resource_attributes
+      @resource_attributes ||= default_attributes
+    end
+
+    def association_columns
+      @association_columns ||= resource_attributes.select{ |key,value| key != value }.values
+    end
+
+    def content_columns
+      @content_columns ||= resource_attributes.select{ |key,value| key == value }.values
```
Comment (timoschilling) [inline lib/active_admin/resource.rb]: `|key, value|`

**Comment by timoschilling** [comment]
Code looks ok to me, but I don't have the time to understand the functionality at the moment.

**Comment by Fivell** [comment]
@timoschilling , when you have time , let's  discuss this, I have an idea about moving this to our [orm folder](https://github.com/activeadmin/activeadmin/tree/master/lib/active_admin/orm/active_record)

what do you think ?

**Comment by Fivell** [comment]
@varyonic , maybe you have time to discuss this?

**Comment by Fivell** [comment]
@deivid-rodriguez , this stuff with remote_const is failed again

**Comment by Fivell** [comment]
@deivid-rodriguez , I do  remember you told me there were a way to fix it

File: lib/active_admin/resource.rb [hunk 2]
```diff
@@ -1,4 +1,6 @@
 require 'active_admin/resource/action_items'
+require 'active_admin/resource/attributes'
+require 'active_admin/resource/ordering'
```
Comment (varyonic) [inline lib/active_admin/resource.rb]: Duplicate require

**Comment by deivid-rodriguez** [comment]
@Fivell Yes, sorry, I forgot about it. Let me add a fix.

File: lib/active_admin/resource.rb (hunk 2 above)
Comment (Fivell) [inline lib/active_admin/resource.rb]: @varyonic , attributes != ordering ... , don't understand what duplication do you mean

Comment (varyonic) [inline lib/active_admin/resource.rb]: ordering is included also L.15

File: lib/active_admin/filters/resource_extension.rb [hunk 3]
```diff
@@ -109,19 +109,11 @@ def filter_lookup
...
       end
 
-      def custom_ransack_filters
-        if resource_class.respond_to?(:_ransackers)
-          resource_class._ransackers.keys.map(&:to_sym)
-        else
-          []
-        end
-      end
-
```
Comment (deivid-rodriguez) [inline lib/active_admin/filters/resource_extension.rb]: This method was not removed, was just moved around. Any reason? Moving it back here would simplify the diff and make reviewing this easier.

File: spec/support/rails_template.rb [hunk 4]
```diff
@@ -29,6 +29,18 @@ class Post < ActiveRecord::Base
...
+      parent.table[:title]
+    end
+
+    ransacker :custom_created_at_searcher do |parent|
+      parent.table[:created_at]
+    end
+
+    ransacker :custom_searcher_numeric, type: :numeric do
+      # nothing to see here
+    end
```
Comment (deivid-rodriguez) [inline spec/support/rails_template.rb]: All this stuff was moved in another PR, I think it's duplicated now.

Comment (Fivell) [inline spec/support/rails_template.rb]: sure you're right

**Comment by Fivell** [comment]
@activeadmin/collaborators ,  if you don't mind I will merge this in evening

**Comment by deivid-rodriguez** [comment]
:+1: I'll try to review it before. If I don't make it, merge away and I'll review post-merge if I feel it's necessary!

**Comment by Fivell** [comment]
@deivid-rodriguez , sure,  all previous comments were fixed and rebased

File: lib/active_admin/csv_builder.rb [hunk 5]
```diff
@@ -105,7 +105,7 @@ def initialize(name, resource = nil, options = {}, block = nil)
 
       def humanize_name(name, resource, humanize_name_option)
         if humanize_name_option
-          name.is_a?(Symbol) && resource.present? ? resource.human_attribute_name(name) : name.to_s.humanize
+          name.is_a?(Symbol) && resource.respond_to?(:resource_class) ? resource.resource_class.human_attribute_name(name) : name.to_s.humanize
```
Comment (deivid-rodriguez) [inline lib/active_admin/csv_builder.rb]: This is the first thing that surprised me, I would expect that we shouldn't need the `respond_to?` check because our resources always respond to `resource_class`. It turns out that the `ResourceDSL` class is initialized with an AA resource object _and_ the underlying object (raw `Post`, for example), and this second object is what's used by the `CSVBuilder`.

Since we are refactoring, I think this dependency in the `ResourceDSL` class is not necessary at all (if it needs to access the underlying resource, it can always do `config.resource_class`). So I would kill the overriden initializer in https://github.com/activeadmin/activeadmin/blob/master/lib/active_admin/resource_dsl.rb#L4-L8 and use `config` instead of `@resource` here: https://github.com/activeadmin/activeadmin/blob/master/lib/active_admin/resource_dsl.rb#L110-L112.

That way we could keep the old `present?` check and just change `resource.human_attribute_name` with `resource.resource_class.human_attribute_name` since this class now gets the actual resource and not its class.

Makes sense?

Comment (Fivell) [inline lib/active_admin/csv_builder.rb]: @deivid-rodriguez , yes, let's try

**Comment by Fivell** [comment]
@deivid-rodriguez , works

File: lib/active_admin/resource/attributes.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
+module ActiveAdmin
+
+  class Resource
+    module Attributes
+
+      def default_attributes
+        return {} unless resource_class.respond_to?(:columns)
```
Comment (deivid-rodriguez) [inline lib/active_admin/resource/attributes.rb]: Can this actually happen?

File: lib/active_admin/views/index_as_table.rb [hunk 7]
```diff
@@ -233,8 +233,8 @@ def default_table
         proc do
           selectable_column
           id_column if resource_class.primary_key
-          resource_class.content_columns.each do |col|
-            column col.name.to_sym
+          active_admin_config.resource_attributes.values.each do |attribute|
```
Comment (deivid-rodriguez) [inline lib/active_admin/views/index_as_table.rb]: You can use `resource_columns` here, no?

Comment (Fivell) [inline lib/active_admin/views/index_as_table.rb]: @deivid-rodriguez , yes

File: lib/active_admin/resource/attributes.rb (hunk 6 above)
Comment (Fivell) [inline lib/active_admin/resource/attributes.rb]: @deivid-rodriguez , don't remember context, maybe for activemodel compatibility or something... looks like it is better to remove this since 
I don't see any test for this to be present

**Comment by Fivell** [comment]
@deivid-rodriguez , fixed and squashed

File: lib/active_admin/resource/attributes.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
+module ActiveAdmin
+
+  class Resource
+    module Attributes
+
+      def default_attributes
+        resource_class.columns.each_with_object({}) do |c, attrs|
+          #skip primary key, STI column and counter cache columns
+          unless  c.name == resource_class.primary_key || c.name == resource_class.inheritance_column || c.name =~ /_count$/
```
Comment (deivid-rodriguez) [inline lib/active_admin/resource/attributes.rb]: I would extract `primary_key_col?` `sti_col?` and `counter_cache_col?` methods here. That way we avoid the need for the comment since the code would be self-explanatory.

File: lib/active_admin/resource/attributes.rb [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+            name = c.name.to_sym
+            attrs[name] = (method_for_column(name) || name)
+          end
+          attrs
```
Comment (deivid-rodriguez) [inline lib/active_admin/resource/attributes.rb]: I think you don't need to explicitly return the accumulator when you use `each_with_object`.

**Comment by Fivell** [comment]
@deivid-rodriguez ,  take a look now

File: lib/active_admin/resource/attributes.rb [hunk 10]
```diff
@@ -0,0 +1,44 @@
+module ActiveAdmin
+
+  class Resource
+    module Attributes
+
+      def default_attributes
+        resource_class.columns.each_with_object({}) do |c, attrs|
+          unless reject_col?(c)
```
Comment (deivid-rodriguez) [inline lib/active_admin/resource/attributes.rb]: Actually, do we need this check at all? I applied the following patch and all features passed:

```diff
diff --git a/lib/active_admin/csv_builder.rb b/lib/active_admin/csv_builder.rb
index ba323302..36fcd8e8 100644
--- a/lib/active_admin/csv_builder.rb
+++ b/lib/active_admin/csv_builder.rb
@@ -21,7 +21,7 @@ module ActiveAdmin
     # resource's content columns
     def self.default_for_resource(resource)
       new resource: resource do
-        column :id
+        # column :id
         resource.content_columns.each { |c| column c }
       end
     end
diff --git a/lib/active_admin/resource/attributes.rb b/lib/active_admin/resource/attributes.rb
index 48063ddf..c587b72d 100644
--- a/lib/active_admin/resource/attributes.rb
+++ b/lib/active_admin/resource/attributes.rb
@@ -24,6 +24,7 @@ module ActiveAdmin
       end
 
       def reject_col?(c)
+        return false
         primary_col?(c) || sti_col?(c) || counter_cache_col?(c)
       end
 
@@ -41,4 +42,4 @@ module ActiveAdmin
 
     end
   end
-end
\ No newline at end of file
+end
diff --git a/lib/active_admin/views/index_as_table.rb b/lib/active_admin/views/index_as_table.rb
index 01323833..5da66c4f 100644
--- a/lib/active_admin/views/index_as_table.rb
+++ b/lib/active_admin/views/index_as_table.rb
@@ -232,7 +232,7 @@ module ActiveAdmin
       def default_table
         proc do
           selectable_column
-          id_column if resource_class.primary_key
+          # id_column if resource_class.primary_key
           active_admin_config.resource_columns.each do |attribute|
             column attribute
           end
```

I can see a point for skipping the inheritance column on STI models (we might want to test this), but I think the `counter_cache` columns can be useful by default, and that we don't really need to skip the primary key.

Comment (Fivell) [inline lib/active_admin/resource/attributes.rb]: This is out of the scope of current PR I'm not going to change functionality please read original message

Comment (deivid-rodriguez) [inline lib/active_admin/resource/attributes.rb]: Sorry @Fivell, I'm not sure I understand. I don't see any counter cache column detection in the previous code. Could you clarify?

Comment (Fivell) [inline lib/active_admin/resource/attributes.rb]: No it not changing functionality counters cache columns and sti columns were always ignored see rails content_columns method implementation

Comment (deivid-rodriguez) [inline lib/active_admin/resource/attributes.rb]: Oh, I didn't know rails had such method, nice. Where are we using it? In `CSVBuilder`, you seem to be using Rails `columns` method, not content columns.

Comment (deivid-rodriguez) [inline lib/active_admin/resource/attributes.rb]: Feel free, to not answer, I'll re-read the PR knowing what I know now. Thanks! :wink:

Comment (deivid-rodriguez) [inline lib/active_admin/resource/attributes.rb]: Ok, I think I get it now.

* _Columns ended with `_id`._ Rails's `content_columns` excludes these, but we don't want to do that because we want to show those in our tables and CSV exports when they are not linked to an association . So we switch from AR `content_columns` to just `columns` and don't filter out these columns. :+1: for that change in behaviour. 

* _Columns ended with `_count`_. By the same argument, I'd say we want to have these columns as well, since they are just regular columns not linked to any association. This would be a change in behavior, just like adding `_id`-ended columns. :+1: for this change in behavior as well. I think it's better behavior and it saves us from duplicating AR functionality. 

* _Primary key column_. In this case, my suggestion doesn't change the behavior but it leads to simpler code because we don't duplicate AR functionality either. So :+1: to simplifying that.

Since this PR is tagged as "refactoring", I'd suggest to try not duplicate AR functionality if we don't need to. So I would:

* Drop `reject_col?` method.
* Drop `counter_cache_col?` method.
* Drop `primary_col?` method.
* Drop https://github.com/activeadmin/activeadmin/blob/master/lib/active_admin/csv_builder.rb#L24 line.
* Drop https://github.com/activeadmin/activeadmin/blob/master/lib/active_admin/views/index_as_table.rb#L235 line.

Comment (Fivell) [inline lib/active_admin/resource/attributes.rb]: @deivid-rodriguez , 

Again I'm not going to change default rendering attributes.
Only one functional change is  that columns ends with _id they are not always foreign keys,  that's one of the reasons why rails content_columns is not suitable, (see foo_id in tests). Also that's why it is not about duplication of AR functionallity. 

Switching CSV export to use columns is out of scope since it is changing behaviour too, columns include also foreign keys and STI columns. 

Columns ended with _count.  They are not regular columns, they are sytem columns to cache counters, they were designed for different things, changing this behaviour anyway is not context of this PR, as well as sti column. User still can ovverride with activeadmin DSL.


Dropping id_column in index_table is bad idea, it is helper method and not just rendering primary key.

Comment (Fivell) [inline lib/active_admin/resource/attributes.rb]: @deivid-rodriguez , feel free to create separate PR with changing default attributes you want to render and we will discuss what kind of columns we should keep and leave and what tests and features we should add to cover this.

Comment (deivid-rodriguez) [inline lib/active_admin/resource/attributes.rb]: * Regarding `_count` columns. Ok, we can consider that the behavior you're adding are bug fixes since some users are complained about it, but this new behavior would be different since nobody has asked for it. So agreed, let's welcome this little extra complexity for now for the sake of keeping the current behavior.

* Regarding the `id_column` method, you're absolutely right, I completely overlooked that. I just applied the little patch and no feature specs broke so I incorrectly assumed it wasn't necessary.

So, agreed, let's not make any of the changes I suggested in https://github.com/activeadmin/activeadmin/pull/4784#discussion_r105568747. I :-1:'d my comment as well! :rofl:

**Comment by deivid-rodriguez** [comment]
Thanks for your patience! :sweat_smile:
